From 6d2f3beca2d05a15287829c2021275c54eedfccc Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <tlanger@maxlinear.com>
Date: Fri, 14 May 2021 16:15:28 +0200
Subject: [PATCH] PONRTSYS-8555: Create new mib_cnt_lock in port-info

and use it in the support functions for mib_counters.
---
 drivers/net/datapath/dpm/datapath.h              | 2 ++
 drivers/net/datapath/dpm/datapath_misc.c         | 1 +
 drivers/net/datapath/dpm/gswip31/datapath_misc.c | 5 ++---
 drivers/net/datapath/dpm/gswip31/datapath_tx.c   | 6 ++----
 4 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index a671f5783cf3..045445928499 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -11,6 +11,7 @@
 #include <linux/netdevice.h>
 #include <linux/platform_device.h>
 #include <linux/atomic.h>
+#include <linux/spinlock.h>
 #include <linux/version.h>
 #include <net/qos_mgr/qos_mgr_tc_hook.h>
 #include <linux/percpu.h>
@@ -688,6 +689,7 @@ struct pmac_port_info {
 	/* These members must be end. */
 	u32 tail;
 	struct dp_subif_info *subif_info;
+	spinlock_t mib_cnt_lock; /* lock for updates from mib_counters module*/
 };
 
 struct ctp_dev {
diff --git a/drivers/net/datapath/dpm/datapath_misc.c b/drivers/net/datapath/dpm/datapath_misc.c
index 13c46b209a12..c52a2d08963f 100644
--- a/drivers/net/datapath/dpm/datapath_misc.c
+++ b/drivers/net/datapath/dpm/datapath_misc.c
@@ -1331,6 +1331,7 @@ u32 alloc_dp_port_subif_info(int inst)
 			return DP_FAILURE;
 		}
 		get_dp_port_info(inst, port_id)->subif_info = sifs;
+		spin_lock_init(&get_dp_port_info(inst, port_id)->mib_cnt_lock);
 	}
 	return DP_SUCCESS;
 }
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index b6d8a4e01cdd..982d0e190173 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -2677,12 +2677,11 @@ int dp_get_reinsert_cnt(int inst, int dp_port, int vap, int flag,
 {
 	struct pmac_port_info *port_info = get_dp_port_info(inst, dp_port);
 	struct dp_subif_info *sif = get_dp_port_subif(port_info, vap);
-	struct hal_priv *priv = HAL(inst);
 
-	DP_TX_CB_LOCK(&priv->cb_lock);
+	spin_lock_bh(&port_info->mib_cnt_lock);
 	memcpy(dp_reins_count, &sif->reins_cnt,	sizeof(*dp_reins_count));
 	DP_DEBUG(DP_DBG_FLAG_DBG, "Reinsert counters\n");
-	DP_TX_CB_UNLOCK(&priv->cb_lock);
+	spin_unlock_bh(&port_info->mib_cnt_lock);
 
 	return DP_SUCCESS;
 }
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_tx.c b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
index 52a93dfc62e3..2050f1de96c7 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_tx.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
@@ -17,13 +17,11 @@ static int dp_reinsert_pkt_cnt(struct pmac_port_info *dp_info, int vap,
 			       struct sk_buff *skb)
 {
 	struct dp_reinsert_count *stats;
-	int inst = 0;
-	struct hal_priv *priv = HAL(inst);
 	int len = skb->len + ETH_FCS_LEN;
 
 	DP_DEBUG(DP_DBG_FLAG_DBG, "len=%d vap=%d\n", len, vap);
 	stats = &get_dp_port_subif(dp_info, vap)->reins_cnt;
-	DP_TX_CB_LOCK(&priv->cb_lock);
+	spin_lock_bh(&dp_info->mib_cnt_lock);
 	if (len <= 64) {
 		stats->dp_64BytePkts++;
 	} else if ((len <= 127) && (len >= 65)) {
@@ -50,7 +48,7 @@ static int dp_reinsert_pkt_cnt(struct pmac_port_info *dp_info, int vap,
 	stats->dp_GoodPkts++;
 	stats->dp_GoodBytes += len;
 
-	DP_TX_CB_UNLOCK(&priv->cb_lock);
+	spin_unlock_bh(&dp_info->mib_cnt_lock);
 
 	return DP_SUCCESS;
 }
